htlcswitch/link.go

```diff
@@ -2127,26 +2127,21 @@ func (l *channelLink) ChanID() lnwire.ChannelID {
 //
 // NOTE: Part of the ChannelLink interface.
 func (l *channelLink) Bandwidth() lnwire.MilliSatoshi {
+	// Get the balance available on the channel for new HTLCs. This takes
+	// the channel reserve into account so HTLCs up to this value won't
+	// violate it.
 	channelBandwidth := l.channel.AvailableBalance()
-	overflowBandwidth := l.overflowQueue.TotalHtlcAmount()
 
 	// To compute the total bandwidth, we'll take the current available
 	// bandwidth, then subtract the overflow bandwidth as we'll eventually
 	// also need to evaluate those HTLC's once space on the commitment
 	// transaction is free.
-	linkBandwidth := channelBandwidth - overflowBandwidth
-
-	// If the channel reserve is greater than the total available balance
-	// of the link, just return 0.
-	reserve := lnwire.NewMSatFromSatoshis(l.channel.LocalChanReserve())
-	if linkBandwidth < reserve {
+	overflowBandwidth := l.overflowQueue.TotalHtlcAmount()
+	if channelBandwidth < overflowBandwidth {
 		return 0
 	}
 
-	// Else the amount that is available to flow through the link at this
-	// point is the available balance minus the reserve amount we are
-	// required to keep as collateral.
-	return linkBandwidth - reserve
+	return channelBandwidth - overflowBandwidth
```

Comment:
`AvailableBalance` will return the balance available after subtracting `htlcCommitFee` (regular commitment fee + fee for adding one extra htlc) and the channel reserve. This means that we already have subtracted the required `1*htlcFee` buffer. So think we should be good?

But you are right, it gets a bit iffy with the overflow queue. The problem is that the overflow queue by definition contains HTLCs of some value, but doesn't account for the fee it would take to add these to the commitment. This calculation wouldn't be an exact science anyway, since for HTLCs from the overflow queue to be added, we first would have to remove some of the other HTLCs.

I think the best way to estimate how much balance we have available at that point would be to subtract the value in the overflow queue plus `len(overflowQueue)*htlcFee`. This will not always be correct, but conservative. wdyt?

Reply:
Maybe you don't need to subtract `len * htlcFee`, because the htlcs can also all be send and settled one after the other. The `htlcFee` will be returned every time. I was thinking to first subtract the sum of the overflow queue and only then see that is left and compare that with `htlcFee`.

Reply:
Discussed offline. It is indeed equivalent

Reply:
One related item is that we don't attempt to manage this overflow queue at all. We assume that things will eventually drain out, so we don't require any sort of active queue management. However, one could argue that if we emptied out this queue (by cancelling the HTLCs backwards) during times of network congestion, then we free up the sending clients (they can retry another link as they assume it's either congestion or a lack of available bandwidth) and also tend to an empty queue. 

There's some active research in this area with papers like the Spider network we could draw from in an attempt to address this when we decide to tackle it. This may also partially mitigate channel jamming attacks as we'll cancel back those HTLCs sooner, which frees up HTLC slots in the _incoming_ channel.